### PR TITLE
feat: support constant folding

### LINKED ISSUE #1268: Code inside `if (false) { ... }` branches (created by plugins replacing env variables) are not removed

### PR DESCRIPTION
**Pull request by hyf0**
### Description

Fixes #1268.

### CODE REVIEW

**Comment by hyf0** [comment]
* **#1401** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1401?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/tests/esbuild/dce/dce_of_destructuring/artifacts.snap [hunk 1]
```diff
@@ -10,12 +10,7 @@ input_file: crates/rolldown/tests/esbuild/dce/dce_of_destructuring
 ```js
 
 //#region entry.js
-var KEEP1 = x;
-var [remove3] = [];
-var [remove4,...remove5] = [...[1, 2], 3];
-var [,,remove6] = [, , 3];
-var [KEEP2] = [x];
-var [KEEP3] = [...{}];
+var remove1, remove2 = null, KEEP1 = x, [remove3] = [], [remove4,...remove5] = [...[1, 2], 3], [,,remove6] = [, , 3], [KEEP2] = [x], [KEEP3] = [...{}], {KEEP4} = {};
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/dce/dce_of_destructuring/artifacts.snap]: `join_vars: false,` is not being respected.

File: crates/rolldown/tests/esbuild/default/export_forms_common_js/artifacts.snap [hunk 2]
```diff
@@ -13,7 +13,7 @@ input_file: crates/rolldown/tests/esbuild/default/export_forms_common_js
 //#region a.js
 var abc;
 var init_a = __esmMin(() => {
-	abc = undefined;
+	abc = void 0;
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/default/export_forms_common_js/artifacts.snap]: This does emit small output. But is it still meaningful in the aspect of bundling with not enabling minifer? If not, we need a option to control this behavior.

File: crates/rolldown/tests/esbuild/default/import_missing_common_js/artifacts.snap [hunk 3]
```diff
@@ -18,9 +18,8 @@ var require_foo = __commonJSMin((exports, module) => {
 //#endregion
 //#region entry.js
 var import_foo = __toESM(require_foo());
-if (false) {
-	console.log((0,import_foo.default)(import_foo.x, import_foo.y));
-}
+if (false) console.log((0,import_foo.default)(import_foo.x, import_foo.y));
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/default/import_missing_common_js/artifacts.snap]: Constant folding is not supported in oxc compersor. Oxc so far only support folding in Ternary operation.

File: crates/rolldown/tests/esbuild/default/require_resolve/artifacts.snap [hunk 4]
```diff
@@ -24,13 +24,12 @@ console.log(require.resolve('@scope/external-pkg'));
 try{
 	console.log(require.resolve('inside-try'));
 }catch(e){
-}if (false) {
-	console.log(require.resolve('dead-code'));
-}
+}if (false) console.log(require.resolve('dead-code'));
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/default/require_resolve/artifacts.snap]: Same comment as https://github.com/rolldown/rolldown/pull/1401/files#r1641953925.

**Comment by hyf0** [comment]
Comments that are required actions have been made. @Boshen cc

File: crates/rolldown/src/utils/pre_process_ast.rs [hunk 5]
```diff
@@ -45,6 +46,22 @@ pub fn pre_process_ast(
     return Err(anyhow::anyhow!("Transform failed, got {:#?}", errors));
   }
 
+  ast.with_mut(|fields| {
+    oxc::minifier::Compressor::new(
+      fields.allocator,
+      CompressOptions {
```
Comment (hyf0) [inline crates/rolldown/src/utils/pre_process_ast.rs]: We expect bundled code look like what it was, so these are all `false`.

**Comment by IWANABETHATGUY** [comment]
According to your comment in https://github.com/rolldown/rolldown/pull/1401/files#r1641953969, the constant folding is not supported yet? Should we closed #1268?

**Comment by hyf0** [comment]
> According to your comment in https://github.com/rolldown/rolldown/pull/1401/files#r1641953969, the constant folding is not supported yet? Should we closed #1268?

This PR won't be merged until oxc supports it. Let me convert it to a draft.

**Comment by Boshen** [comment]
Let me try try.

**Review by Boshen — APPROVED** [review]
Big achievement!

**Comment by IWANABETHATGUY** [comment]
> Big achievement!

let's go!!!

**Pull request merged**